src/dict.js

```diff
@@ -69,68 +74,64 @@ Sk.builtin.dict.tp$call = function(args, kw) {
 Sk.abstr.setUpInheritance("dict", Sk.builtin.dict, Sk.builtin.object);
 Sk.abstr.markUnhashable(Sk.builtin.dict);
 
-var kf = Sk.builtin.hash;
-
-Sk.builtin.dict.prototype.key$lookup = function (bucket, key) {
-    var item;
-    var eq;
-    var i;
-
-    // Fast path: We spend a *lot* of time looking up strings
-    // in dictionaries. (Every attribute access, for starters.)
-    if (key.ob$type === Sk.builtin.str) {
-        for (i = 0; i < bucket.items.length; i++) {
-            item = bucket.items[i];
-            if (item.lhs.ob$type === Sk.builtin.str && item.lhs.v === key.v) {
-                return item;
-            }
-        }
-        return null;
+function kf(key) {
+    let key_hash;
+    if (key.$savedKeyHash_) {
+        return key.$savedKeyHash_;
     }
-
-    for (i = 0; i < bucket.items.length; i++) {
-        item = bucket.items[i];
-        eq = Sk.misceval.richCompareBool(item.lhs, key, "Eq");
-        if (eq) {
-            return item;
-        }
+    if (Sk.builtin.checkString(key)) {
+        key_hash = "_" + key.$jsstr();
+        key.$savedKeyHash_ = key_hash;
+        return key_hash;
+    } else if (typeof key === "string") {
+        return "_" + key;
```

Comment:
It might be better to return an unmodified string here, to creating a new string and possibly take advantage of previous string canonicalisation for the hash lookup; you could return an object literal (smth like `{key:key_hash}`) for the non-string case.

Reply:
do you mind providing a little more of a code snippet.

I think I follow but just in case I don't...

worth noting that I used the leading underscore since javascript objects only preserver insertion order for strings so if the user had a `{'1': 'foo'}` then without the leading underscore the insertion order would be lost...

and also 86-87 are only used by `sysModules` if #1098 gets merged these two lines will go... 
technically i could remove them already but it makes `sysModules` a little faster and avoids sending `strings` to `Sk.misceval.richCompareBool` which is only designed for `skulpt objects`

Reply:
I was imagining something like:

```js
if (key.ob$type === Sk.builtin.str) {
    return key.$jsstr();
}
return { key: Sk.builtin.hash(key).v };
```

which is used as

```js
key = kf(obj);
if (typeof key == "string") {
    return entries[key];
}
return entries[key.key]
```

Dunno how well a non-string saved key would work with the rest of the system.

The numbers thing is a good point -- I would suggest maybe special-casing prefixes? You could e.g. do:

```js
function sanitize_key(key) {
  return key
}

if (key.ob$type === Sk.builtin.str) {
    return key.$jsstr().replace(/^[0-9!#_]/, "!$&");
}
return "#" + Sk.builtin.hash(key).v;
```

which would make sure that only non-string hashes start with a `#`, and no hashes start with a number or a `_` (for avoiding conflicts with setting `__proto__`).

Reply:
I've had another play around. I think the `"_"` proceeding keys is the easiest solution... you never know when the replacing might conflict with some random string...? and it fixes the problem with ints in javascript not preserving insertion order...

The  `"_" + key.$jsstr()` should only happen the first time a string is entered into a dictionary. The next time that key is looked up `$savedKeyHash_` will return the appropriate value.

for the non string keys, I now just return the integer `hash value`. Which seems cleaner.

thoughts?

Reply:
Your call -- my regex approach didn't replace anything, but rather reduced the number of cases where it prepended `!` to only those where the string starts with one of `0-9!#_` (i.e., the "danger cases"). Always prepending something also works, albeit less efficiently. At least don't use `_` though, or you'll accidentally set the prototype on the object if your string is `_proto__`.

Reply:
Oh i see now! my regex needs some revision! will have a think about this one tomorrow thanks!

Reply:
I like @LeszekSwirski's idea, and I can see how this could be a performance win, but it looks like we've got the potential for collisions here:

> ```js
> key = kf(obj);
> if (typeof key == "string") {
>     return entries[key];
> }
> return entries[key.key]
> ```

In particular, if we store the string `foo` and then a Python object that _hashes_ to `foo`, won't they overwrite each other in `entries`? This was the whole purpose of the prefixing...

Reply:
Re the other point - if we _are_ prefixing, it does seem like `_` is a poor choice of character, because of the existence of `__proto__`! Perhaps we should use `!` or something else that is definitely not going to be in a JS identifier...

Reply:
er, sorry, I think I meant "buckets" on the second return. Either way, I think adding a prefix to the "complex" hash and making sure that the "simple" hash doesn't start with that prefix is the better solution than a wrapper object. 

Regarding `Object.create(null)` vs `{__proto__: null}` -- there's no difference in the objects they produce :) setting __proto__ on _any_ object will set its prototype.

Reply:
> In particular, if we store the string "foo" and a Python object that hashes to "foo", won't they overwrite each other in entries?

This won't happen... hash values from the python hash algorithm are always integers. 

So with the current version the hash value from `kf` will return a js string for python str ie the fast path. Or a JavaScript number if it's not a string (the .v of the return value from `builtin.hash`). 

I'll update this algorithm - on reflection I'll go with @LeszekSwirski solution of prepending based on the regex idea. Like you said it's efficient, avoids the `str(1)` not preserving insertion order as well as conflicts with the complex keys in entries (like `#2_0` - resulting from inserting an item with hash value 2 into the zeroth index of a bucket)

(Nb skulpt doesn't currently allow for user defined `__hash__` functions and it it did the builtin hash function would have to check it was a suitable return value (size and type)). 

Re the `__proto__` @albertjan I'll need you check if this affects interned!!

Reply:
updated! more suggestions welcome... I think it's there now...